sympy/solvers/solveset.py

```diff
@@ -998,6 +1003,163 @@ def _solveset(f, symbol, domain, _check=False):
+
+    if a.is_Add:
+        # g + h = a
+        g, h = a.as_independent(symbol)
+        if g is not S.Zero:
+            return _invert_modular(Mod(h, m), (rhs - Mod(g, m)) % m, symbol)
+
+    if a.is_Mul:
+        # g*h = a
+        g, h = a.as_independent(symbol)
+        if g is not S.One:
+            return _invert_modular(Mod(h, m), (rhs*invert(g, m)) % m, symbol)
+
+    if a.is_Pow:
+        # base**expo = a
+        base, expo = a.args
+        if expo.has(symbol) and not base.has(symbol):
+            return expo, FiniteSet(discrete_log(m, rhs, a.base))
+
+    return a, ImageSet(Lambda(n, m*n + rhs % m), S.Integers)
+
+
+def _solve_modular(f, symbol, domain):
+    r"""
+    Helper function for solving modular equations of type ``A - Mod(B, C) = 0``,
+    where A can or cannot be a function of symbol, B is surely a function of
+    symbol and C is an integer.
+
+    Currently only cases when A is not a function of symbol is been implemented.
+
+    Parameters
+    ==========
+
+    f : Expr
+        The modular equation to be solved, ``f = 0``
+
+    symbol : Symbol
+        The variable in the equation to be solved.
+
+    domain : Set
+        A set over which the equation is solved.
+
+    Returns
+    =======
+
+    A set of solutions satisfying the given modular equation.
+    A ``ConditionSet`` if the equation is unsolvable.
+
+    Examples
+    ========
+
+    >>> from sympy.solvers.solveset import _solve_modular as solve_modulo
+    >>> from sympy import S, Symbol, sin
+    >>> from sympy.core.mod import Mod
+    >>> x = Symbol('x')
+    >>> solve_modulo(Mod(5*x - 8, 7) - 3, x, S.Reals)
+    ImageSet(Lambda(_n, 7*_n + 5), Integers)
+    >>> solve_modulo(Mod(x**2 - 2*x + 1, 7) - 3, x, S.Complexes)
+    ImageSet(Lambda(_n, {1 - sqrt(7*_n + 3), sqrt(7*_n + 3) + 1}), Integers)
+    >>> solve_modulo(Mod(sin(x), 7) - 3, x, S.Reals) # not solvable
+    ConditionSet(x, Eq(Mod(sin(x), 7) - 3, 0), Reals)
+    """
+    # extract modterm and g_ys from f
+    modterm = list(f.atoms(Mod))[0]
+    t = Dummy('t', integer=True)
+    f_t = f.xreplace({modterm: t})
+    g_ys = _solveset(f_t, t, S.Integers)
+    result = EmptySet()
+
+    # partial inversion using _invert_modular.
+    if isinstance(g_ys, FiniteSet):
+        for rhs in g_ys:
+            # Case: A-> function of symbol, can be extended here in future.
+            # Currently handling Case: A -> not a function of symbol.
+            if not rhs.has(symbol):
+                f_x, y = _invert_modular(modterm, rhs, symbol)
+                # solution finding
+                if f_x is symbol:
+                    result += y
+                elif isinstance(y, ImageSet):
+                    lamda_expr = y.lamda.expr
+                    lamda_vars = y.lamda.variables
+                    sol_set = _solveset(f_x - lamda_expr, symbol, S.Complexes)
+                    result += ImageSet(Lambda(lamda_vars, sol_set), S.Integers)
+                elif isinstance(y, FiniteSet):
```

Comment:
The purpose of this and the above `elif` isn't very clear. Can you throw some light on why these solving is done again? I actually expect `_invert_modular` to provide with all the solutions.